File: speechify-assessment-prep/src/ts/parser.ts
// Stack-based parser scaffold that turns tokens into a node tree.
import type { Token } from "./tokenizer";

export interface Node {
  tag: string;
  attrs: Record<string, string>;
  children: Node[];
  text: string;
}

export function parseNodes(tokens: Token[]): Node[] {
  const root: Node = { tag: "root", attrs: {}, children: [], text: "" };
  const stack: Node[] = [root];

  for (const token of tokens) {
    if (token.kind === "OPEN") {
      const node: Node = { tag: token.tag, attrs: token.attrs, children: [], text: "" };
      stack[stack.length - 1].children.push(node);
      stack.push(node);
    } else if (token.kind === "SELF") {
      const node: Node = { tag: token.tag, attrs: token.attrs, children: [], text: "" };
      stack[stack.length - 1].children.push(node);
    } else if (token.kind === "TEXT") {
      stack[stack.length - 1].text += token.text;
    } else if (token.kind === "CLOSE") {
      // TODO: assert stack top matches token.tag before popping.
      stack.pop();
    } else {
      const _exhaustive: never = token;
      throw new Error(`Unhandled token ${_exhaustive}`);
    }
  }

  // TODO: assert stack.length === 1 for well-formed input.
  return root.children;
}

File: speechify-assessment-prep/src/py/parser_.py
"""Stack-based parser scaffold that turns tokens into a node tree."""
from __future__ import annotations

from typing import Dict, List, Tuple

Token = Tuple[str, ...]
Node = Dict[str, object]


def parse_nodes(tokens: List[Token]) -> List[Node]:
    """Build a simple node tree from the token stream."""
    root: Node = {"tag": "root", "attrs": {}, "children": [], "text": ""}
    stack: List[Node] = [root]
    for tk in tokens:
        kind = tk[0]
        if kind == "OPEN":
            tag, attrs = tk[1], tk[2]
            node: Node = {"tag": tag, "attrs": attrs, "children": [], "text": ""}
            stack[-1]["children"].append(node)
            stack.append(node)
        elif kind == "SELF":
            tag, attrs = tk[1], tk[2]
            node = {"tag": tag, "attrs": attrs, "children": [], "text": ""}
            stack[-1]["children"].append(node)
        elif kind == "TEXT":
            stack[-1]["text"] += tk[1]
        elif kind == "CLOSE":
            tag = tk[1]
            # TODO: assert or validate that stack[-1]["tag"] == tag before popping.
            stack.pop()
        else:
            raise ValueError(f"Unknown token kind: {kind}")
    # TODO: assert len(stack) == 1 to confirm proper closure.
    return root["children"]  # type: ignore[return-value]
